rptest/direct_consumer_tests leadership transfer test & fixes

joe-redpanda opened the pull request:
Adds test which repeatedly and randomly transfers leadership in the
background

Fixes for leadership transfer related problems
Issues fixed:
1. mutex lock update direct_consumer::update_fetchers
   - fast leadership transfers launched concurrent update_fetchers
     executions. interleaved execution mangled the fetch_offsets value
     for the parition and resulted in a reset to 0
2. skip adding to the data_queue if the fetch offset was not updated
   - when maybe_update_fetch_offset returns false, it indicates an
     assignment change. Namely for quick A->B->A broken transfers this
     resulted redundant offsets.
     - for the ABA case, the next fetch will supercede the old fetch
   - resolution is to discard the old fetch
3. newly assigned partitions should always be a part of the next fetch
   request
4. a partition should only be removed from the next fetch request if
    the fetcher assignment epochs are the same, this avoids A->B->A
    reassignment race conditions
5. a partition should be removed from the 'to_forget' index
    if has been reassigned to the fetcher
6. its possible for a request to receive a response which was
    previously part of its assignment set if a 'forget' has not yet been
    sent. Treat this case the same as 'changed assignment epoch'

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Bug Fixes

* fixes leadership transfer related problems in direct consumer

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -764,7 +769,7 @@ ss::future<> fetcher::assign_partition(
       .fetch_offset = offset,
       .assignment_epoch = next_epoch(),
-    };
+      .incremental_include = true};
```
good eye 🙏

oleiman commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -153,7 +154,12 @@ ss::future<> direct_consumer::start() {
+
+    // close the gate and break the lock together, no need to drain all
+    // operations
+    auto f = _gate.close();
+    _subscriptions_lock.broken();
+    co_await std::move(f);
```
does creating the `gate::close` future before breaking the semaphore actually have any effect? I don't think we'll yield until the `co_await`, so I'm not sure I understand the significance here.

joe-redpanda replied:
I did it just so things shut down faster

joe-redpanda replied:
If it looks weird we can probably just drop it

dotnwat replied:
> I did it just so things shut down faster

why do you believe that it would result in faster shutdown? are you thinking at like the instruction order level, or at the scheduling level?

joe-redpanda replied:
I had no strong reason to write it that way

My thought was maybe we could close the gate to stop callers from getting to the mutex?
then the mutex closes
then we drain the gate

90% sure that you're right in that it doesn't matter
No caller is making progress until we yield the cpu which only happens when we await the gate close operation anyway

so either a fiber will see the gate and mutex magically become closed while the fiber was suspended
or the fiber will see the gate and mutex magically become closed while the fiber was suspended

dotnwat replied:
offline discussion: no semantic difference, and no scheduling changes

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -529,10 +529,15 @@ fetcher::process_fetch_response(
-                if (updated_offset) {
-                    dirty_partitions[topic_data.topic].insert(
-                      part_data.partition_id);
+                if (!updated_offset) {
+                    // case when partition is either
+                    // 1. partition is not assigned to this fetcher
+                    // 2. assignment epoch changed
+                    // for either skip
+                    continue;
                 }
+                dirty_partitions[topic_data.topic].insert(
+                  part_data.partition_id);
```
looks equivalent to me. is the change mainly to add a place for the comment?

joe-redpanda replied:
two brokers A&B:
if leadership gets kicked A->B->A quickly, we can
                                           ^ start a fetch request here
                                                        ^ and complete it here
At which point the fetch ends up on the queue because it **is assigned here** just not in the same epoch

Which causes duplicate batches to appear in the output queue because a future fetch will be sent with the same start offset

offset not updated maps afaik 1 to 1 with "don't put on the queue"

joe-redpanda replied:
the 'continue' skips putting the resultant fetch on the queue

joe-redpanda replied:
dropped

oleiman replied:
for posterity: the code is still present but it was my mistake reading block boundaries.

oleiman commented on tests/rptest/direct_consumer_tests/direct_consumer_test.py:
```diff
@@ -24,15 +31,58 @@
+    @cluster(num_nodes=5)
     def test_basic_consuming_from_topic(self):
         topic_name = "test-topic"
+        #msg_count = 2000000
```
dead code?

oleiman commented on tests/rptest/direct_consumer_tests/direct_consumer_test.py:
```diff
@@ -111,12 +165,20 @@ def get_consumption():
+            troublemaker.stop()
+            producer.stop()
             verifier.stop()
+            producer.wait(timeout_sec=600)
```
nitpick: is this necessary? if we consumed `msg_count`, then the producer should be done w/ it's work, right?

joe-redpanda replied:
probably not necessary. Saw elsewhere for letting KGO shutdown more gracefully which hopefully would save me some manual env cleanup

oleiman replied:
kk. yeah if it doesn't act weirdly with an already stopped producer no issue from me. does seem like a noop though

oleiman reviewed (commented):
few minor comments/questions, but generally lgtm

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70748-->
<!--41cbaf2e-72cf-4f32-9725-352f5e4e6fc9-->
<!--8110966d-753d-4923-b04c-f7b1b816a4ea-->
<!--title-->
## Retry command for [Build#70748](https://buildkite.com/redpanda/redpanda/builds/70748)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/direct_consumer_tests/direct_consumer_test.py::DirectConsumerVerifierTest.test_basic_consuming_from_topic
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--be2cb0d2-e746-4ee7-b2c0-30509b87febb-->
<!--TEST RESULTS-->
<!--f74c2550-b1a9-4506-ae39-6b9bdfafe8bc-->
<!--TEST RESULTS-->
<!--84572e3b-6185-436b-89cd-10261576cb33-->
<!--TEST RESULTS-->
<!--f12f865c-8e40-4125-9ad7-1bdb89244e09-->
<!--TEST RESULTS-->
<!--cd9967db-9a7e-4590-8225-ecdcc44ac1a8-->
<!--TEST RESULTS-->
<!--3940ee97-66f7-41ca-8c84-14d30085b4f7-->
<!--TEST RESULTS-->
<!--24488991-e0e8-4fd2-a137-36760ad127c8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70748">build#70748</a></summary>


| test_class                 | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:---------------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| DirectConsumerVerifierTest | test_basic_consuming_from_topic | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70748#0198a538-6837-40e5-805a-82d5a9375818 | FAIL          | 0/21     | The test has failed across all retries |
| DirectConsumerVerifierTest | test_basic_consuming_from_topic | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70748#0198a53f-30cd-4267-80ac-14f7321e47d7 | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70772">build#70772</a></summary>


| test_class                 | test_method                              | test_arguments          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:-----------------------------------------|:------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DirectConsumerVerifierTest | test_basic_consuming_from_topic          | null                    | integration | https://buildkite.com/redpanda/redpanda/builds/70772#0198a5ee-deef-45ca-8670-5645bebe5307 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| DirectConsumerVerifierTest | test_basic_consuming_from_topic          | null                    | integration | https://buildkite.com/redpanda/redpanda/builds/70772#0198a5f5-e46f-42cb-b09a-fda1f956af77 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 |
| FeaturesMultiNodeTest      | test_license_upload_and_query            | null                    | integration | https://buildkite.com/redpanda/redpanda/builds/70772#0198a5ee-def4-4837-a969-47782067d27c | FLAKY         | 18/21    | upstream reliability is '92.6829268292683'. current run reliability is '85.71428571428571'. drift is 6.96864 and the allowed drift is set to 50. The test should PASS  |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/70772#0198a5f5-e473-4a83-8ea8-e498835eb97f | FLAKY         | 20/21    | upstream reliability is '99.42528735632183'. current run reliability is '95.23809523809523'. drift is 4.18719 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70828">build#70828</a></summary>


| test_class                 | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:---------------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| DirectConsumerVerifierTest | test_basic_consuming_from_topic | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70828#0198aa11-4386-4853-8fda-ccb3ff61b6c2 | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70838">build#70838</a></summary>


| test_class                 | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:---------------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| DirectConsumerVerifierTest | test_basic_consuming_from_topic | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70838#0198aa98-fe72-400a-b421-b4f2c4fbed58 | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70886">build#70886</a></summary>


| test_class                        | test_method                          | test_arguments             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------------|:-------------------------------------|:---------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets          | {"versions_to_upgrade": 3} | integration | https://buildkite.com/redpanda/redpanda/builds/70886#0198b00a-a9f2-449a-8980-d9a137389e8f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| ProducersAdminAPITest             | test_producers_state_api_during_load | null                       | integration | https://buildkite.com/redpanda/redpanda/builds/70886#0198b013-6b55-4072-9c70-fbb704f78470 | FLAKY         | 20/21    | upstream reliability is '99.72527472527473'. current run reliability is '95.23809523809523'. drift is 4.48718 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71007">build#71007</a></summary>


| test_class                 | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:---------------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| DirectConsumerVerifierTest | test_basic_consuming_from_topic | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71007#0198c3bd-c45c-49ae-a72b-3519be7dd5eb | FAIL          | 0/21     | The test has failed across all retries |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71129">build#71129</a></summary>


| test_class                | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write                           | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71129#0198cacb-615e-4fe5-b26b-3343c87ca15e | FLAKY         | 19/21    | upstream reliability is '93.10344827586206'. current run reliability is '90.47619047619048'. drift is 2.62726 and the allowed drift is set to 50. The test should PASS |
| ProducersAdminAPITest     | test_producers_state_api_during_load | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71129#0198cacb-6159-426f-9a30-aee77c726f17 | FLAKY         | 20/21    | upstream reliability is '99.73118279569893'. current run reliability is '95.23809523809523'. drift is 4.49309 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -153,7 +154,12 @@ ss::future<> direct_consumer::start() {
 ss::future<> direct_consumer::stop() {
     _cluster->unregister_metadata_cb(_metadata_callback_id);
     _fetched_data_queue->stop();
-    co_await _gate.close();
+
+    // close the gate and break the lock together, no need to drain all
+    // operations
+    auto f = _gate.close();
+    _subscriptions_lock.broken();
+    co_await std::move(f);
```
i think that this

```
    auto f = _gate.close();
    _subscriptions_lock.broken();
    co_await std::move(f);
```

should be functionality equivalent to

```
_subscriptions_lock.broken();
co_await _gate.close();
```

because in both cases the breaking of the semaphore and the setting of the closed flag in the gate happen without yielding to the reactor.

can you provide a description of the race you believe is happening? something like the following is usually sufficient to illustrate

```
F1: <statements executed>
scheduling point
F2: <statements executed>
scheduling point
F1: resumes with <statements...>
etc...
```

joe-redpanda replied:
No race or problem. The only motivation was 'shut down faster'

swapping

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70772-->
<!--d6e8545e-efeb-4695-bf52-f6183faeadc9-->
<!--fe96eb04-31ee-4ee5-b9e2-f11c5d09fee7-->
<!--title-->
## Retry command for [Build#70772](https://buildkite.com/redpanda/redpanda/builds/70772)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/direct_consumer_tests/direct_consumer_test.py::DirectConsumerVerifierTest.test_basic_consuming_from_topic
```
<!--/body-->

mmaslankaprv commented on tests/rptest/direct_consumer_tests/direct_consumer_test.py:
```diff
@@ -111,12 +164,20 @@ def get_consumption():
                 f"Expected {msg_count} messages, got {final_state.total_consumed_messages}"
 
+            assert int(final_state.non_monotonic_fetches) == 0, \
+                f"Non-monotomic fetches found, number of nm fetches: {final_state.non_monotonic_fetches}"
```
nit: typo

joe-redpanda replied:
Took me longer than I'd like to admit to figure out which word was spelled wrong

mmaslankaprv commented:
great work, Joe would you mind separating the three fixes, each into separate commit ?

joe-redpanda committed:
direct_consumer_test.py: add leadership transfer

Test improvement: adds a background thread to drive random leadership
transfers while direct consumer consumes.

joe-redpanda committed:
direct_consumer: update race cond fix

add a mutex lock on direct_consumer::update_fetchers
 - fast leadership transfers launched concurrent update_fetchers
   executions. interleaved execution mangled the fetch_offsets value
   for the parition and resulted in a reset to 0

joe-redpanda committed:
direct_consumer/fetcher: dedupe batches

skip adding to the data_queue if the fetch offset was not updated
 - if we decline to update the fetch offset (maybe_update_fetch_offset)
   we should not offer the fetched data to the ouptut queue
   - case topic unassigned: the data is no longer needed
   - case partition unassigned: the data is no longer needed
   - case assignment epoch changed: a future fetch will gather the same
     batches, this fetch is a duplicate and should be skipped

joe-redpanda committed:
direct_consumer/fetcher: include new parititons

bugfix on incremental fetch: newly assigned partitions should always be
part of the next fetch

joe-redpanda commented:
> great work, Joe would you mind separating the three fixes, each into separate commit ?

Split everything up and updated the commit messages accordingly. Let me know if theres something I should document better

oleiman commented on tests/rptest/direct_consumer_tests/direct_consumer_test.py:
```diff
@@ -9,9 +9,16 @@
 
 from typing import Any
 
+import time
```
nitpick: if this test would fail without subsequent changes, I would suggest putting it on the other end of the stack to preserve `git bisect`ability

joe-redpanda replied:
fixed

oleiman commented on src/v/kafka/client/direct_consumer/tests/direct_consumer_fixture_test.cc:
```diff
@@ -175,6 +175,29 @@ class consumer_fixture : public cluster_test_fixture {
+                break;
+            }
+        }
+        vassert(
+          current_leader != original_leader,
+          "never detected leadership change, test precondition failed");
```
nitpick: since the this function returns void, you should be able to use gtest asserts here.

oleiman commented on src/v/kafka/client/direct_consumer/tests/direct_consumer_fixture_test.cc:
```diff
@@ -175,6 +175,29 @@ class consumer_fixture : public cluster_test_fixture {
+        const auto original_leader = this->get_leader(ntp);
+        auto current_leader = original_leader;
+
+        this->shuffle_leadership(ntp).get();
```
nitpick: do you really need the `this` pointer here?

oleiman commented on tests/rptest/direct_consumer_tests/direct_consumer_test.py:
```diff
@@ -24,13 +31,55 @@
+                            args=(topic_spec, ))
+        return thread
+
+    @cluster(num_nodes=5)
```
thought: I wonder if we'd benefit from having separate test cases with & without forcing leadership transfer. Dunno how interesting the stable leadership case is in general.

joe-redpanda replied:
I'm going to add parameterization in the follow on. I've been concurrently working on random node failures.
I'll switch this to troublemakers: {None, ShuffleLeadership, RandomFailures}

oleiman reviewed (dismissed):
few nits but lgtm

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70828-->
<!--60dcef1e-6dcb-4f52-865b-84f04aa359f4-->
<!--title-->
## Retry command for [Build#70828](https://buildkite.com/redpanda/redpanda/builds/70828)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/direct_consumer_tests/direct_consumer_test.py::DirectConsumerVerifierTest.test_basic_consuming_from_topic
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/70838-->
<!--b02d4df3-c958-4b54-a2b2-0753b1be9af4-->
<!--title-->
## Retry command for [Build#70838](https://buildkite.com/redpanda/redpanda/builds/70838)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/direct_consumer_tests/direct_consumer_test.py::DirectConsumerVerifierTest.test_basic_consuming_from_topic
```
<!--/body-->

joe-redpanda commented:
Converting to draft while I investigate. Tests pass consistently locally, but are routinely missing batches here.

joe-redpanda committed:
direct_consumer/fetcher: fix fetcher state race

Leadership transfers from broker A -> broker B -> broker A
sometimes occur fast enough that a fetch from an earlier assignment
epoch completes after reassignment to A. If the old epoch fetch has no
updates for the transferred partition, it will be discluded from the
next fetch inappropriately.

The resolution is to check epoch before modifying the fetcher state for
a given assignment.

joe-redpanda commented:
There are more race conditions namely in the fast ABA leadership transfer case.
For the sake of getting fixes for ci, I've removed the integration test update that stresses leadership transfers.

joe-redpanda committed:
direct_consumer/fetcher: aba race condition fix

In the case of sufficiently fast ABA leadership transfers between
brokers, the fetcher may have a partition both being added and
forgotten. In this case, assign_parition should remove the partition
from those to be forgotten.

variable rename for readability

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -565,20 +565,43 @@ fetcher::process_fetch_response(
-                auto p_it = ps.find(p.partition_id);
-                if (p_it != ps.end() && !partition_err) {
-                    p_it->second.incremental_include = false;
+                auto partition_iterator = partition_map.find(p.partition_id);
```
change makes sense. good catch. I'm wondering whether we should do the update under the lock regardless. not 100% sure that would eliminate the need for an epoch check though.

joe-redpanda replied:
I think the epoch check will always be needed so long as we're allowing multi fiber updates because we launch a fetch with a snapshot of state and allow for updates while its in flight

I feel like the clean thing to do is have getters/setters and demand an epoch for comparison on the setter.

alternatively, I'm wondering if we could have assign and unassign simply submit updates to a queue of updates that gets drained at the start of each fetcher loop, that way the user of the state (the fetch loop) has a much better handle on how and when updates occur.

oleiman replied:
> we launch a fetch with a snapshot of state and allow for updates while its in flight

right, yeah good point.

> getters/setters and demand an epoch for comparison

sounds legit.

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -794,6 +794,27 @@ ss::future<> fetcher::assign_partition(
+    // in the case of fast leadership transfers, we may have a partition both
+    // being added and forgotten, in which case, make sure that it is only being
+    // added
+    auto forget_topic_iterator = _topics_to_partitions_to_forget.find(tp.topic);
```
makes me wonder whether it's the best way to organize this bit of state. e.g. maybe 'forgotten' should just be another field on fetcher_state. much easier to enforce mutual exclusivity with 'added'.

iirc, I organized it this way to avoid erasing from `_partitions` in `process_fetch_response`, but we're mutating it anyway to flip the include flag. anyway food for thought.

oleiman reviewed (dismissed):
lgtm. thanks for fixing all of those.

just to confirm: you left fetch sessions enabled by default, but the collection of fixes means leadership xfer tests will consistently pass, right?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/71007-->
<!--fcb6030d-fe9b-4f1a-bbc8-60a239cdf8b3-->
<!--4ab72665-a52c-4aa3-a91f-e55583d9f544-->
<!--title-->
## Retry command for [Build#71007](https://buildkite.com/redpanda/redpanda/builds/71007)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/direct_consumer_tests/direct_consumer_test.py::DirectConsumerVerifierTest.test_basic_consuming_from_topic
```
<!--/body-->

joe-redpanda commented:
> lgtm. thanks for fixing all of those.
> 
> just to confirm: you left fetch sessions enabled by default, but the collection of fixes means leadership xfer tests will consistently pass, right?

Yeah my stats are 50/50 PASS locally

I know of one additional issue which hits ~1/100 but I'm capping this PR.
I'll fix that with random node operations fixes

CI failed on the vtools issue (I'm working on it), I'll merge once thats fixed and I can showcase a 50x ci repeat

joe-redpanda committed:
direct_consumer/fetcher: unbidden response fix

In fast leadership changes, it is possible for a fetcher to receive a
response for a partition that was not assigned when it snapshots the
partition epochs.

This happens in the following circumstance:
1. reassignment of paritition P from A -> B
2. fetch begins and snapshots epochs
3. reassignment of P from B -> A
4. produce to P
5. A responds with updates for P to fetcher for A

This was causing an illegal memory access because the snapped epochs for
the request would be missing a value for P.

The resolution is to check for not only different epochs but to allow
for absense of epoch in the snapshotted request epochs.

The behavior is functionally the same as an epoch change.

joe-redpanda committed:
direct_consumer/application: better logging

Adds logging for duplicated offsets. This is not illegal because this
can and will be caught by write_at_offset_stm but instances of it should
be logged.

joe-redpanda commented:
/ci-repeat 3 tests/rptest/direct_consumer_tests/direct_consumer_test.py::DirectConsumerVerifierTest
dt-repeat=50

joe-redpanda commented:
/ci-repeat 10
release
dt-repeat 5
skip-redpanda-build
tests/rptest/direct_consumer_tests/direct_consumer_test.py

joe-redpanda commented:
/ci-repeat 5
sanitizer
dt-repeat 10
tests/rptest/direct_consumer_tests/direct_consumer_test.py

mmaslankaprv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -274,7 +274,7 @@ bool fetcher::maybe_update_fetch_offset(
   kafka::offset last_received,
   kafka::offset high_watermark,
-  assignment_epoch assignment_epoch) {
+  std::optional<assignment_epoch> maybe_response_offset) {
```
should that be epoch ?

joe-redpanda replied:
whoops, definitely

mmaslankaprv commented on src/v/kafka/client/direct_consumer/tests/BUILD:
```diff
@@ -41,6 +41,7 @@ redpanda_cc_gtest(
     srcs = [
         "direct_consumer_fixture_test.cc",
     ],
+    cpu = 1,
```
this change is unrelated

mmaslankaprv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -565,20 +565,43 @@ fetcher::process_fetch_response(
+                    if (current_epoch == fetched_epoch) {
+                        partition_iterator->second.incremental_include = false;
+                    } else {
+                        logger().trace(
```
use `vlog` macro

mmaslankaprv commented on src/v/kafka/client/direct_consumer/fetcher.h:
```diff
@@ -263,7 +263,7 @@ class fetcher {
     fetch_session_state _session_state;
     topic_partition_map<partition_fetch_state> _partitions;
-    topic_partition_map<model::partition_id> _partitions_to_forget;
+    topic_partition_map<model::partition_id> _topics_to_partitions_to_forget;
```
name change seems unrelated to the commit, can it be a separate one ?

joe-redpanda replied:
thats fair. I think i'll make a general renames commit to help with readability

we have
_partitions and _partitions_to_forget which then both get loops that expand into 
topic_it, parition_it, auto& paritition = parition_it->second and so on

joe-redpanda replied:
kicking renames to a follow on

joe-redpanda committed:
direct_consumer_fixture_test: leadership transfer

Adds a leadership transfer tests to direct_consumer_fixture_test.

Snaps the fixture test to a single cpu s.t. cluster_fixture utilities
work.

joe-redpanda committed:
direct_consumer/verifier: application logging

Add logs on the offsets which have been fetched.

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -565,20 +565,44 @@ fetcher::process_fetch_response(
+                        partition_iterator->second.incremental_include = false;
+                    } else {
+                        vlog(
+                          logger().trace,
```
nit: `DEBUG`?

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -795,6 +795,27 @@ ss::future<> fetcher::assign_partition(
+    // in the case of fast leadership transfers, we may have a partition both
+    // being added and forgotten, in which case, make sure that it is only being
+    // added
+    auto forget_topic_iterator = _partitions_to_forget.find(tp.topic);
+    if (forget_topic_iterator != _partitions_to_forget.end()) {
+        // topic is in the 'to forget map', now is the partition in the topic's
+        // map
+        auto& partitions_to_forget = forget_topic_iterator->second;
+        auto forget_partition_iterator = partitions_to_forget.find(
+          tp.partition);
+        if (forget_partition_iterator != partitions_to_forget.end()) {
+            // the ntp is both being added and forgotten, make sure its only
+            // added
+            partitions_to_forget.erase(forget_partition_iterator);
+            if (partitions_to_forget.empty()) {
+                // cleanup if this was the last partition
+                _partitions_to_forget.erase(forget_topic_iterator);
+            }
+        }
+    }
```
thought: feels like we've written this `lookup in A -> lookup in B-> erase from B -> check B empty -> erase B from A` sequence several times. is there a generic(ish) helper function to be had? ok if not, just an observation.

joe-redpanda replied:
100% agree, I will take these cleanups later because we're bleeding on these issues right now.

my plan is something like
```cpp
template<typename E>
maybe_lookup(map<topic, map<partition, E>>& nested_map, topic topic, partition, partition)
-> optional<something?>
```

1. we often need both of the iterators to erase the outer map if we empty out the inner map
2. iterators aren't fun because you can't simply call iterator.is_end, you instead check against the original collection
3. references and optionals don't play well together

I'll give it more thought

oleiman replied:
sounds fine to me. if Michal is green on this PR I'm happy

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -678,78 +718,94 @@ ss::future<kafka::error_code> fetcher::maybe_initialise_fetch_offsets(
-    auto offsets = co_await do_list_offsets(std::move(req));
-    if (offsets.has_error()) {
-        if (is_retriable_error(offsets.error())) {
+    auto list_offsets_response = co_await do_list_offsets(std::move(req));
```
it's hard to see whether anything changed in this section of the diff...could we isolate the renaming in its own commit and mark it as such in the commit message?

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -586,12 +613,25 @@ fetcher::process_fetch_response(
-                      = epochs.find(topic)->second.find(p.partition_id)->second;
 
-                    if (current_epoch == fetched_epoch) {
+                    auto maybe_request_epoch = find_assignment_epoch(
+                      topic, p.partition_id, epochs);
+
+                    if (!maybe_request_epoch) {
+                        // see maybe_update_fetch_offset for explanation
+                        vlog(
+                          logger().info,
+                          "unbidden response on ntp: {}/{}",
+                          topic,
+                          p.partition_id);
+                        continue;
+                    }
+
+                    const auto request_epoch = maybe_request_epoch.value();
+
+                    if (assigned_epoch == request_epoch) {
```
nitpick: is this not equivalent to 

```suggestion
                    auto maybe_request_epoch = find_assignment_epoch(
                      topic, p.partition_id, epochs);
                    if (assigned_epoch == maybe_request_epoch) {
```

which is to say, do we really care about having a different log message for these two cases? functionally a mismatch is a mismatch, right?

joe-redpanda replied:
I didn't realize optional offered an operator==. I think we care if BOTH are missing but I can vassert on that

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -678,78 +718,94 @@ ss::future<kafka::error_code> fetcher::maybe_initialise_fetch_offsets(
+
+            const auto assigned_epoch = p_it->second.assignment_epoch;
+
+            auto maybe_response_epoch = find_assignment_epoch(
+              response_topic.topic, response_partition.partition_id, epochs);
+
+            if (!maybe_response_epoch) {
+                vlog(
+                  logger().warn,
+                  "[broker: {} received a list topics response which was not "
+                  "requested on ntp: {}/{}",
+                  _id,
+                  response_topic.topic,
+                  response_partition.partition_id);
+                continue;
+            }
+
+            const auto request_epoch = maybe_response_epoch.value();
+
+            if (request_epoch != assigned_epoch) {
```
similar comment with regard to logging and processing the mismatch. if the mismatch is a natural and recoverable side effect of async assign/unassign, then does it really need a separate warn log?

joe-redpanda replied:
this should not be possible for list topics which is why this one is warn

oleiman commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -678,78 +718,94 @@ ss::future<kafka::error_code> fetcher::maybe_initialise_fetch_offsets(
+                  response_partition.offset,
                   request_epoch,
                   p_it->second.assignment_epoch);
+                continue;
```
just to confirm, you're fixing a bug here yeah?

oleiman commented on src/v/kafka/client/direct_consumer/tests/direct_consumer_fixture_test.cc:
```diff
@@ -80,6 +80,46 @@ TEST_P(BasicConsumerFixture, TestBasicConsumption) {
       model::offset(1019));
 }
 
+TEST_P(BasicConsumerFixture, TestBasicLeadershipTransfer) {
```
nitpick: if this test would flake before subsequent changes, might consider putting it at the top of the stack

joe-redpanda replied:
we're good, this only flakes on aba changes which this does not do.
I didn't attempt to recreate time dependent issues in this PR like aba leadership transfers.

I will likely borrow the idea of having a function that runs a singular loop s.t. we can get more deterministic fixture tests in the future

oleiman replied:
👌

oleiman reviewed (commented):
couple marginal nits but overall looking great.

joe-redpanda commented:
@oleiman I agree with your comments. @bharathv currently needs this in dev to progress on his tests so I'm going to ingest your comments as the first commit of the follow on
